components/script/dom/elementinternals.rs

```diff
@@ -0,0 +1,371 @@
+                }
+                self.validation_anchor.set(Some(a));
+            },
+        }
+        Ok(())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-validationmessage>
+    fn GetValidationMessage(&self) -> Fallible<DOMString> {
+        // This check isn't in the spec but it's in WPT tests and it maintains
+        // consistency with other methods that do specify it
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.validation_message.borrow().clone())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-validity>
+    fn GetValidity(&self) -> Fallible<DomRoot<ValidityState>> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.validity_state())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-labels>
+    fn GetLabels(&self) -> Fallible<DomRoot<NodeList>> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.labels_node_list.or_init(|| {
+            NodeList::new_labels_list(
+                self.target_element.upcast::<Node>().owner_doc().window(),
+                &*self.target_element,
+            )
+        }))
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-willvalidate>
+    fn GetWillValidate(&self) -> Fallible<bool> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.is_instance_validatable())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-form>
+    fn GetForm(&self) -> Fallible<Option<DomRoot<HTMLFormElement>>> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.form_owner.get())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-checkvalidity>
+    fn CheckValidity(&self) -> Fallible<bool> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.check_validity())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-reportvalidity>
+    fn ReportValidity(&self) -> Fallible<bool> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.report_validity())
+    }
+}
+
+fn submission_value_from(value: &Option<FileOrUSVStringOrFormData>) -> SubmissionValue {
+    match value {
+        None => SubmissionValue::None,
+        Some(FileOrUSVStringOrFormData::File(f)) => SubmissionValue::File(DomRoot::from_ref(f)),
+        Some(FileOrUSVStringOrFormData::USVString(s)) => SubmissionValue::USVString(s.clone()),
+        Some(FileOrUSVStringOrFormData::FormData(fd)) => SubmissionValue::FormData(fd.datums()),
```

Comment:
Avoid using "extreme" abbreviations. This should probably be "form_data":

```suggestion
        Some(FileOrUSVStringOrFormData::FormData(form_data)) => SubmissionValue::FormData(form_data.datums()),
```